xpm/xpm.go

```diff
@@ -0,0 +1,126 @@
+package icon
+
+import (
+	"bufio"
+	"fmt"
+	"image"
+	"image/color"
+	"io"
+	"strconv"
+	"strings"
+)
+
+func parseXPM(data io.Reader) (image.Image, error) {
+	var colCount, charSize int
+	colors := make(map[string]color.Color)
+	var img *image.NRGBA
+
+	rowNum := 0
+	scan := bufio.NewScanner(data)
+	for scan.Scan() {
+		row := scan.Text()
+		if row == "" || row[0] != '"' {
+			continue
+		}
+		row = stripQuotes(row)
+
+		if rowNum == 0 {
+			w, h, cols, size := parseDimensions(row)
+			img = image.NewNRGBA(image.Rectangle{image.Point{}, image.Point{w, h}})
+			colCount = cols
+			charSize = size
+		} else if rowNum <= colCount {
+			id, c, err := parseColor(row, charSize)
+			if err != nil {
+				return nil, err
+			}
+
+			if id != "" {
+				colors[id] = c
+			}
+		} else {
+			parsePixels(row, charSize, rowNum-colCount-1, colors, img)
+		}
+		rowNum++
+	}
+	return img, scan.Err()
+}
+
+func parseColor(data string, charSize int) (id string, c color.Color, err error) {
+	if len(data) == 0 {
+		return
+	}
+	parts := strings.Fields(data)
+	if len(parts) == 2 && parts[0] == "c" {
+		parts = []string{" ", "c", parts[1]}
+	} else if len(parts) != 3 {
+		return
+	} else if parts[1] != "c" {
+		return
+	}
+
+	color, err := stringToColor(parts[2])
+	return data[:charSize], color, nil
+}
+
+func parseDimensions(data string) (w, h, i, j int) {
+	if len(data) == 0 {
+		return
+	}
+	parts := strings.Split(data, " ")
+	if len(parts) != 4 {
+		return
+	}
+
+	w, _ = strconv.Atoi(parts[0])
```

Comment:
This one also can return an error, right?

Reply:
That one never even crossed my mind, added